File: paper_algo.cpp
//-----------------Tawkir Aziz Rahman - Optimized with Original Logic---------------------//

#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

// Optimized data structures to replace sets
class FastEdgeSet
{
private:
    vector<vector<bool>> adj;
    vector<pair<ll, ll>> edges;
    ll n;

public:
    FastEdgeSet(ll size) : n(size)
    {
        adj.assign(size, vector<bool>(size, false));
    }

    void insert(pair<ll, ll> edge)
    {
        ll u = min(edge.first, edge.second);
        ll v = max(edge.first, edge.second);
        if (!adj[u][v])
        {
            adj[u][v] = true;
            edges.push_back({u, v});
        }
    }

    void erase(pair<ll, ll> edge)
    {
        ll u = min(edge.first, edge.second);
        ll v = max(edge.first, edge.second);
        if (adj[u][v])
        {
            adj[u][v] = false;
            edges.erase(remove(edges.begin(), edges.end(), make_pair(u, v)), edges.end());
        }
    }

    bool find(pair<ll, ll> edge)
    {
        ll u = min(edge.first, edge.second);
        ll v = max(edge.first, edge.second);
        return u < n && v < n && adj[u][v];
    }

    vector<pair<ll, ll>> &getEdges() { return edges; }

    FastEdgeSet copy()
    {
        FastEdgeSet result(n);
        result.edges = edges;
        result.adj = adj;
        return result;
    }
};

pair<ll, ll> make_pair_ordered(ll first, ll second)
{
    return make_pair(min(first, second), max(first, second));
}

bool checkLeftmostBlockingChord(ll root, ll n, FastEdgeSet &chords, pair<ll, ll> chord)
{
    ll currentBlockedChord = n - 1;
    for (ll i = n - 2; i >= 0; i--)
    {
        if (!chords.find(make_pair_ordered(root, i)))
        {
            break;
        }
        else
        {
            currentBlockedChord = i;
        }
    }

    vector<pair<ll, ll>> leftBlockingChords;
    for (auto &a : chords.getEdges())
    {
        if (a.second == currentBlockedChord && a.first > root)
        {
            leftBlockingChords.push_back(a);
        }
    }

    if (leftBlockingChords.empty())
        return false;

    pair<ll, ll> leftmostChord = *min_element(leftBlockingChords.begin(), leftBlockingChords.end());
    return (leftmostChord.first == chord.first && leftmostChord.second == chord.second);
}

pair<ll, ll> flipChord(ll root, ll n, FastEdgeSet &chords, FastEdgeSet &allPairs, pair<ll, ll> chord)
{
    vector<ll> assocOfFirst, assocOfSecond;

    // Find neighbors of chord.first
    for (auto &a : allPairs.getEdges())
    {
        if (a.first == chord.first)
        {
            assocOfFirst.push_back(a.second);
        }
        if (a.second == chord.first)
        {
            assocOfFirst.push_back(a.first);
        }
    }

    // Find neighbors of chord.second
    for (auto &a : allPairs.getEdges())
    {
        if (a.first == chord.second)
        {
            assocOfSecond.push_back(a.second);
        }
        if (a.second == chord.second)
        {
            assocOfSecond.push_back(a.first);
        }
    }

    vector<ll> commons;
    sort(assocOfFirst.begin(), assocOfFirst.end());
    sort(assocOfSecond.begin(), assocOfSecond.end());

    set_intersection(assocOfFirst.begin(), assocOfFirst.end(),
                     assocOfSecond.begin(), assocOfSecond.end(),
                     back_inserter(commons));

    chords.erase(chord);
    allPairs.erase(chord);

    pair<ll, ll> newChord = make_pair_ordered(commons[0], commons[1]);
    chords.insert(newChord);
    allPairs.insert(newChord);

    return newChord;
}

void generateAllTriangulations(ll root, ll n, FastEdgeSet allPairs, FastEdgeSet chords,
                               vector<vector<pair<ll, ll>>> &answers)
{
    if (n < 3)
        return;
    if (n == 3)
        return;

    // Add current triangulation
    answers.push_back(chords.getEdges());

    // Try flipping each chord
    auto currentChords = chords.getEdges();
    for (auto &chord : currentChords)
    {
        FastEdgeSet allPairsCopy = allPairs.copy();
        FastEdgeSet chordsCopy = chords.copy();

        auto newChord = flipChord(root, n, chordsCopy, allPairsCopy, chord);

        bool isLeftmostBlockingChord = checkLeftmostBlockingChord(root, n, chordsCopy, newChord);
        if (isLeftmostBlockingChord)
        {
            generateAllTriangulations(root, n, allPairsCopy, chordsCopy, answers);
        }
    }
}

vector<vector<pair<ll, ll>>> generateTriangulationsForFace(vector<ll> &face)
{
    ll n = face.size();
    FastEdgeSet allPairs(n);
    FastEdgeSet chords(n);
    vector<vector<pair<ll, ll>>> answers;

    // Add boundary edges
    for (ll i = 0; i < n; i++)
    {
        allPairs.insert(make_pair_ordered(i, (i + 1) % n));
    }

    // Add initial fan triangulation
    for (ll i = 2; i < n - 1; i++)
    {
        chords.insert(make_pair_ordered(0, i));
        allPairs.insert(make_pair_ordered(0, i));
    }

    generateAllTriangulations(0, n, allPairs, chords, answers);

    // Map indices to actual face vertices
    vector<vector<pair<ll, ll>>> result;
    for (const auto &triangulation : answers)
    {
        vector<pair<ll, ll>> mappedTriangulation;
        for (const auto &edge : triangulation)
        {
            mappedTriangulation.push_back(make_pair_ordered(face[edge.first], face[edge.second]));
        }
        result.push_back(mappedTriangulation);
    }

    return result;
}

// Fast combination generator
void generateAllCombinations(vector<vector<vector<pair<ll, ll>>>> &allTriangulations,
                             vector<vector<pair<ll, ll>>> &allCombinations)
{
    vector<ll> indices(allTriangulations.size(), 0);

    while (true)
    {
        vector<pair<ll, ll>> currentCombination;
        for (size_t i = 0; i < allTriangulations.size(); i++)
        {
            for (const auto &chord : allTriangulations[i][indices[i]])
            {
                currentCombination.push_back(chord);
            }
        }
        allCombinations.push_back(currentCombination);

        // Generate next combination
        ll pos = allTriangulations.size() - 1;
        while (pos >= 0)
        {
            indices[pos]++;
            if (indices[pos] < (ll)allTriangulations[pos].size())
            {
                break;
            }
            indices[pos] = 0;
            pos--;
        }

        if (pos < 0)
            break;
    }
}

void printAllTriangulations(const vector<vector<pair<ll, ll>>> &answers)
{
    cout << "Total triangulations found: " << answers.size() << endl;
    cout << "Printing all triangulations" << endl;
    for (size_t i = 0; i < answers.size(); i++)
    {
        cout << "Triangulation " << i + 1 << ": { ";
        for (const auto &edge : answers[i])
        {
            cout << "( " << edge.first << " , " << edge.second << " ) ";
        }
        cout << "}" << endl;
    }
}

vector<vector<pair<ll, ll>>> eradicateMultiEdge(vector<vector<pair<ll, ll>>> &allCombinations)
{
    vector<vector<pair<ll, ll>>> tempCombinations;
    if (allCombinations.empty())
        return tempCombinations;

    ll standard = allCombinations[0].size();
    for (auto &combination : allCombinations)
    {
        set<pair<ll, ll>> uniqueEdges;
        for (auto &p : combination)
        {
            uniqueEdges.insert(make_pair_ordered(p.first, p.second));
        }
        if (uniqueEdges.size() == standard)
        {
            vector<pair<ll, ll>> cleanCombination(uniqueEdges.begin(), uniqueEdges.end());
            tempCombinations.push_back(cleanCombination);
        }
    }
    return tempCombinations;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<vector<ll>> faces;
    faces.push_back({1, 2, 3, 4, 5, 6});

    vector<vector<vector<pair<ll, ll>>>> allTriangulations;
    for (auto &face : faces)
    {
        vector<vector<pair<ll, ll>>> triangulations = generateTriangulationsForFace(face);
        cout << "Generated triangulations for face of size " << face.size()
             << ": " << triangulations.size() << " triangulations" << endl;
        allTriangulations.push_back(triangulations);
    }

    cout << "\nTotal triangulation groups: " << allTriangulations.size() << endl;
    for (size_t i = 0; i < allTriangulations.size(); i++)
    {
        cout << "Face " << i + 1 << ": " << allTriangulations[i].size() << " triangulations" << endl;
    }

    vector<vector<pair<ll, ll>>> allCombinations;
    generateAllCombinations(allTriangulations, allCombinations);
    cout << "Total combinations generated: " << allCombinations.size() << endl;

    allCombinations = eradicateMultiEdge(allCombinations);

    // Sort for consistency
    for (auto &combination : allCombinations)
    {
        sort(combination.begin(), combination.end());
    }
    sort(allCombinations.begin(), allCombinations.end());

    // Remove duplicates
    auto last = unique(allCombinations.begin(), allCombinations.end());
    allCombinations.erase(last, allCombinations.end());

    cout << "Final unique combinations: " << allCombinations.size() << endl;

    // Print results if reasonable size
    if (allCombinations.size() <= 50)
    {
        printAllTriangulations(allCombinations);
    }

    return 0;
}

File: paper_algo.py
import sys
from typing import List, Set, Tuple, Dict
from copy import deepcopy

# Global counters
geenralTriangulationsCall = 0
findLeftMostBlockingChordCall = 0
flipChordCall = 0

def makepair(a: int, b: int) -> Tuple[int, int]:
    return (a, b) if a < b else (b, a)

def printVector(v: List, st: int = -1, fin: int = -1) -> None:
    if st == -1:
        st = 0
    if fin == -1:
        fin = len(v)
    print(' '.join(map(str, v[st:fin])))
    
def printVectorPair(v: List[Tuple], st: int = -1, fin: int = -1) -> None:
    if st == -1:
        st = 0
    if fin == -1:
        fin = len(v)
    print(' '.join([f"({x[0]},{x[1]})" for x in v[st:fin]]))
    
def printMatrix(v: List[List], st: int = -1, fin: int = -1) -> None:
    if st == -1:
        st = 0
    if fin == -1:
        fin = len(v)
    for row in v[st:fin]:
        printVector(row)
        
def printStructure(t) -> None:
    print(' '.join(map(str, t)))
    
def printAllPairs(s: Set[Tuple[int, int]]) -> None:
    print(' '.join([f"({x[0]},{x[1]})" for x in s]]))
    
def copySetOfPairs(src: Set[Tuple[int, int]]) -> Set[Tuple[int, int]]:
    return {makepair(x[0], x[1]) for x in src}

def checkLeftmostBlockingChord(root: int, n: int, chords: Set[Tuple[int, int]], chord: Tuple[int, int]]) -> bool:
    global findLeftMostBlockingChordCall
    findLeftMostBlockingChordCall += 1
    
    def normalizeVertex(vertex: int) -> int:
        return (vertex - root) % n
    
    normalizedChords = set()
    for c in chords:
        first_norm = normalizeVertex(c[0])
        second_norm = normalizeVertex(c[1])
        normalizedChords.add(makepair(first_norm, second_norm))
    
    chord_first_norm = normalizeVertex(chord[0])
    chord_second_norm = normalizeVertex(chord[1])
    normalizedChord = makepair(chord_first_norm, chord_second_norm)
    
    currentBlockedChord = n - 1
    for i in range(n - 2, -1, -1):
        if makepair(0, i) not in normalizedChords:
            break
        else:
            currentBlockedChord = i
    
    leftBlockingChords = set()
    for a in normalizedChords:
        if a[1] == currentBlockedChord and a[0] > 0:
            leftBlockingChords.add(a)
    
    if not leftBlockingChords:
        return False
    
    leftmostChord = min(leftBlockingChords)
    return leftmostChord[0] == normalizedChord[0] and leftmostChord[1] == normalizedChord[1]

def flipChord(root: int, n: int, chords: Set[Tuple[int, int]], allPairs: Set[Tuple[int, int]], chord: Tuple[int, int]]) -> Tuple[int, int]:
    global flipChordCall
    flipChordCall += 1
    
    assocOfFirst = set()
    assocOfSecond = set()
    
    for a in allPairs:
        if a[0] == chord[0]:
            assocOfFirst.add(a[1])
        if a[1] == chord[0]:
            assocOfFirst.add(a[0])
    
    for a in allPairs:
        if a[0] == chord[1]:
            assocOfSecond.add(a[1])
        if a[1] == chord[1]:
            assocOfSecond.add(a[0])
    
    commons = []
    for a in assocOfFirst:
        if a in assocOfSecond:
            commons.append(a)
    
    chords.discard(chord)
    allPairs.discard(chord)
    new_chord = makepair(commons[0], commons[1])
    chords.add(new_chord)
    allPairs.add(new_chord)
    return new_chord

def generateAllTriangulations(root: int, n: int, allPairs: Set[Tuple[int, int]], chords: Set[Tuple[int, int]], answers: Set[Tuple[Tuple[int, int], ...]]) -> None:
    global geenralTriangulationsCall
    geenralTriangulationsCall += 1
    
    if n < 3:
        print("Invalid input for triangulation")
        return
    
    if n == 3:
        print("Base case: Triangulation of a triangle")
        return
    
    answers.add(frozenset(chords))
    
    for a in list(chords):
        allPairsCopy = copySetOfPairs(allPairs)
        chordsCopy = copySetOfPairs(chords)
        pair = flipChord(root, n, chordsCopy, allPairsCopy, a)
        isLeftmostBlockingChord = checkLeftmostBlockingChord(root, n, chordsCopy, pair)
        
        if isLeftmostBlockingChord:
            generateAllTriangulations(root, n, allPairsCopy, chordsCopy, answers)

def printAllTriangulations(answers: Set[Tuple[Tuple[int, int], ...]]) -> None:
    print(f"Total triangulations found: {len(answers)}")

def findHighestSafeRoot(n: int, innerchords: Set[Tuple[int, int]]) -> int:
    s = set(range(n))
    for a in innerchords:
        s.discard(a[0])
        s.discard(a[1])
    return max(s)

def generateOuterTriangulationsFromAnInnerTriangulation(root: int, n: int, allPairs: Set[Tuple[int, int]], chords: Set[Tuple[int, int]], innerChords: Set[Tuple[int, int]], answers: Set[Tuple[Tuple[int, int], ...]]) -> None:
    if n < 3:
        print("Invalid input for triangulation")
        return
    
    if n == 3:
        print("Base case: Triangulation of a triangle")
        return
    
    minChord = min(chords) if chords else (float('inf'), float('inf'))
    minInnerChord = min(innerChords) if innerChords else (float('inf'), float('inf'))
    
    if minChord < minInnerChord:
        return
    
    allChords = chords.union(innerChords)
    rightlength = (n - 3) * 2
    if len(allChords) < rightlength:
        return
    
    merged = chords.union(innerChords)
    answers.add(frozenset(merged))
    
    for a in list(chords):
        allPairsCopy = copySetOfPairs(allPairs)
        chordsCopy = copySetOfPairs(chords)
        pair = flipChord(root, n, chordsCopy, allPairsCopy, a)
        isLeftmostBlockingChord = checkLeftmostBlockingChord(root, n, chordsCopy, pair)
        
        if isLeftmostBlockingChord:
            generateOuterTriangulationsFromAnInnerTriangulation(root, n, allPairsCopy, chordsCopy, innerChords, answers)

def generateAllTriangulationsForBiconnectedGraph(root: int, n: int, allPairs: Set[Tuple[int, int]], chords: Set[Tuple[int, int]], answers: Set[Tuple[Tuple[int, int], ...]]) -> None:
    tempAnswers = set()
    innerChords = set()
    tempAllPairs = copySetOfPairs(allPairs)
    
    for i in range(2, n - 1):
        innerChords.add(makepair(0, i))
        tempAllPairs.add(makepair(0, i))
    
    generateAllTriangulations(root, n, tempAllPairs, innerChords, tempAnswers)
    
    for a in tempAnswers:
        innerChordsCopy = copySetOfPairs(a)
        initChords = set()
        tempAllPairs = copySetOfPairs(allPairs)
        highestSafeRoot = findHighestSafeRoot(n, innerChordsCopy)
        
        for i in range(2, n - 1):
            initChords.add(makepair(highestSafeRoot, (highestSafeRoot + i) % n))
            tempAllPairs.add(makepair(highestSafeRoot, (highestSafeRoot + i) % n))
        
        temptempAnswers = set()
        generateOuterTriangulationsFromAnInnerTriangulation(highestSafeRoot, n, tempAllPairs, initChords, innerChordsCopy, temptempAnswers)
        
        for b in temptempAnswers:
            answers.add(b)

def generateViaMergingAlgorithm(n: int, mergedAnswers: Set[Tuple[Tuple[int, int], ...]]) -> None:
    allPairs = set()
    chords = set()
    answers = set()
    
    for i in range(n):
        allPairs.add(makepair(i, (i + 1) % n))
    
    for i in range(2, n - 1):
        chords.add(makepair(0, i))
        allPairs.add(makepair(0, i))
    
    generateAllTriangulations(0, n, allPairs, chords, answers)
    
    answers_list = list(answers)
    for i in range(len(answers_list)):
        for j in range(len(answers_list)):
            if i == j:
                continue
            a = answers_list[i]
            b = answers_list[j]
            has_common = False
            for c in a:
                if c in b:
                    has_common = True
                    break
            if not has_common:
                merged = set(a).union(set(b))
                mergedAnswers.add(frozenset(merged))

def compareSet(a: Set[Tuple[int, int]], b: Set[Tuple[int, int]]) -> bool:
    if len(a) != len(b):
        return False
    return a == b

def compareTriangulations(a: Set[Tuple[Tuple[int, int], ...]], b: Set[Tuple[Tuple[int, int], ...]]) -> bool:
    if len(a) != len(b):
        return False
    a_sorted = sorted([sorted(list(x)) for x in a])
    b_sorted = sorted([sorted(list(x)) for x in b])
    return a_sorted == b_sorted

def calculate(n: int) -> None:
    global geenralTriangulationsCall, findLeftMostBlockingChordCall, flipChordCall
    
    allPairs = set()
    chords = set()
    answers = set()
    
    for i in range(n):
        allPairs.add(makepair(i, (i + 1) % n))
    
    mergedAnswers = set()
    generateViaMergingAlgorithm(n, mergedAnswers)
    
    geenralTriangulationsCall = 0
    findLeftMostBlockingChordCall = 0
    flipChordCall = 0
    
    generateAllTriangulationsForBiconnectedGraph(0, n, allPairs, chords, answers)
    
    bl = compareTriangulations(answers, mergedAnswers)
    if bl:
        print(f"For n={n}, sizes ={len(answers)}=={len(mergedAnswers)}")
    else:
        print(f"For n={n}, comparison failed")

def main():
    for i in range(4, 11):
        calculate(i)
        print(f"geenralTriangulationsCall: {geenralTriangulationsCall}")
        print(f"findLeftMostBlockingChordCall: {findLeftMostBlockingChordCall}")
        print(f"flipChordCall: {flipChordCall}")
        print("----------------------------------------")

if __name__ == "__main__":
    main()